### DISCUSSION TITLE
Effectively stopping a process in good time.

### DISCUSSION THREAD

**Discussion by cmaughan** [DaanDeMeyer/reproc#53] [Q&A]
I'm still a little confused about stopping a long running process.
Suppose I have an internal mechanism to stop it, or it will simply finish its work 'soon'.

proc.stop(... wait : 10s, terminate :5s, kill 5s)

Question 1: If I call this function, and after 2 seconds the process naturally exits, my understanding here is that I will still wait for 10 seconds.  So how to 'exit early' if the process has cleanly closed, otherwise wait for 10s?  i.e. I don't want to wait if I don't have to.

Question 2: if the 10 seconds expire, then the process is sent a SIGTERM after 5 more seconds, and then a SIGKILL after 5 more seconds?

**Comment by daandemeyer** [comment]
The wait will stop early if the process exits early.

The stop actions work as follows: The action specifies what to do, the timeout specifies how long to wait before going to the next action (or return if it's the last action). So your example will first wait max 10 seconds for the process to exit on its own, if it hasn't exited after 10 seconds, reproc will send SIGTERM and wait max 5 more seconds for the process to exit. If it hasn't exited after 5 seconds, reproc will send SIGKILL and wait max 5 more seconds for the process to exit. If it exits within 5 seconds, reproc returns the exit status, otherwise, reproc will return an error.

**Reply by cmaughan** [reply]
Unfortunately that's not what I'm seeing.  A wait of 10s will always block for 10s , even though in the process manager the process has gone.  That's on windows at least...

Ah, the SIGTERM is 'how long to wait for the sigterm, not how long till sigterm'; got it ;)

**Reply by daandemeyer** [reply]
I'll try and reproduce later today

**Reply by cmaughan** [reply]
It looks to me like WSAPoll is blocking on the timeout (called from proc_wait, pipe_poll)

If I did proc->wait(0s) , is there an easy way to see if the process has actually exited so I don't need to continue?  There are a lot of possible error codes returned; not sure how to be sure it has gone.

**Reply by cmaughan** [reply]
I think in my scenario, the process I started has already stopped before I call wait, but the block in WSAPoll happens anyway.
I tried _not_ stopping the process, and proc->wait(0s) returns 0/success without a timeout; so I'm not sure even how to know what state it is in.
I don't know if there is anything peculiar about it (it is a ruby.exe process which runs indefinitely).

**Reply by daandemeyer** [reply]
Could you add some logging in `reproc_wait` and `pipe_poll` to see which values are being returned by WSAPoll and WSAGetLastError? That would help in figuring out what might be going wrong.

**Reply by cmaughan** [reply]
I added some quick logging, the results make no sense to me.  I check that the server is running, get the pid.  Check it in the process monitor.

Each wait blocks for the full 5 seconds in WSAPoll, returning 0.  The second wait is waiting even though the process with the same pid has gone.

```
[API] [DEBUG] : Waiting for ruby server: 26412
[API] [INFO ] : Wait before Exit 5 seconds (process->stop with wait 5s)

WSAPoll: 0
[API] [INFO ] : Result: 0 message: The operation completed successfully.

(Process 26412 is still active)

[API] [DEBUG] : SendOSC/exit (causes the process to exit)

[API] [INFO ] : Wait after Exit 5 seconds:

WSAPoll: 0
[API] [INFO ] : Result: 0 message: The operation completed successfully.
```

**Reply by cmaughan** [reply]
Additional (same as above, but calling wait(), then stop() each with 5s delay, then quitting the process and doing it again:

```
[API] [DEBUG] : Waiting for ruby server: 1056

[API] [INFO ] : Stop (5s) Result: 0 message: The operation completed successfully.
[API] [INFO ] : Wait (5s) Result: -258 message: The wait operation timed out.

[API] [INFO ] : Asking server process to exit...

[API] [INFO ] : Stop (5s) Result: 0 message: The operation completed successfully.
[API] [INFO ] : Wait (5s) Result: -258 message: The wait operation timed out.
```

(All WSAPoll == 0, all calls blocked for 5s.  Process not running after the middle exit)

**Reply by cmaughan** [reply]
... and checking the windows way:
```
auto hProcess = OpenProcess(SYNCHRONIZE, TRUE, pid);
auto res = WaitForSingleObject(hProcess, milliseconds);
```
yields results I would expect - WAIT_TIMEOUT before the process has exited, and WAIT_OBJECT_0 afterwards (i.e. the process has gone).  This is with the same pid I get from proc->id().
So I think there is something wrong here on the windows side....

**Reply by daandemeyer** [reply]
So reproc detects exits of the child process by having it inherit a pipe (or a socket in the Windows case) and waiting for the pipe to be closed by calling WSAPoll() with events set to POLLIN. Since that's not happening here, I'm wondering if something the ruby process is doing causes the socket to not get closed correctly.

**Reply by cmaughan** [reply]
That is entirely possible I guess.  I will try to look into the ruby server side tonight.  FYI; this is an API layer I'm writing for SonicPi; and the ruby process does use sockets when sending OSC messages among other things; it is also managing a couple of other processes.

**Reply by daandemeyer** [reply]
>it is also managing a couple of other processes.

This might be a possible cause of the problem. If the child processes inherit the exit pipe and don't exit immediately, reproc won't be able to notice the parent process has exited.

**Reply by daandemeyer** [reply]
Can you try again with the lasest commit on the main branch? It's a long shot but I made some changes that might fix the issue. At the very least, calling `reproc_wait` with `REPROC_INFINITE` as the timeout should now return immediately when the child process exits.

**Reply by cmaughan** [reply]
Sorry for the delay; I plan on digging into this tomorrow; I'll try your update.

**Reply by daandemeyer** [reply]
I had to remove it unfortunately as it broke other use cases. I would suggest looking into whether the child processes spawned by the server are doing anything fishy. If they inherit all pipes from the server process and outlive it, then there's a good chance that's the cause of the issue.

(Unfortunately, you're in a bit of uncharted territory here, I haven't yet tested reproc with processes that spawn other processes themselves)

**Reply by cmaughan** [reply]
I noticed on windows at least that you can specify a pipe is not inheritable when created. Would that help?

**Reply by daandemeyer** [reply]
Yes and reproc uses that internally, but some pipes have to be inherited by the child process, including one pipe which we use to figure out if the process has exited. Unfortunately, there's no way to say that a pipe should only be inherited by a direct child process and not by any processes spawned by that child process.
